sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -7,9 +7,214 @@
+              @total_recorded_attributes += 1
+            end
+          end
+          self
+        end
+
+        # Add an Event to Span
+        #
+        # Note that the OpenTelemetry project
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic-conventions.md
+        # documents} certain "standard event names and keys" which have
+        # prescribed semantic meanings.
+        #
+        # @param [String, Callable] name_or_event_formatter The name of the event
+        #   or an EventFormatter, a lazily evaluated callable that returns an
+        #   Event instance.
+        # @param [optional Hash<String, Object>] attrs One or more key:value pairs, where
+        #   the keys must be strings and the values may be string, boolean or
+        #   numeric type. This argument should only be used when passing in a
+        #   name, not an EventFormatter.
+        # @param [Time] timestamp optional timestamp for the event.
+        #
+        # @return [self] returns itself
+        def add_event(name_or_event_formatter, attrs = nil, timestamp: nil)
+          super
+          timed_event =
+            if name_or_event_formatter.instance_of?(String)
+              TimedEvent.new(name_or_event_formatter, attrs, timestamp || Time.now)
+            else
+              name_or_event_formatter.call
+            end
+          @mutex.synchronize do
+            if @ended
+              # TODO: logger.log(Logger::DEBUG, 'Calling add_event on an ended Span.')
+            else
+              @events ||= []
+              @events << timed_event
+              @events.shift if @events.size > trace_config.max_events_count
+              @total_recorded_events += 1
+            end
+          end
+          self
+        end
+
+        # Adds a link to another Span from this Span. The linked Span can be from
+        # the same or different trace. See {OpenTelemetry::Trace::Link} for a description.
+        #
+        # @param [SpanContext, Callable] span_context_or_link_formatter The
+        #   SpanContext context of the Span to link with this Span or a
+        #   LinkFormatter, a lazily evaluated callable that returns a Link
+        #   instance.
+        # @param [optional Hash<String, Object>] attrs Map of attributes associated with
+        #   this link. Attributes are key:value pairs where key is a string and
+        #   value is one of string, boolean or numeric type. This argument should
+        #   only be used when passing in a SpanContext, not a LinkFormatter.
+        #
+        # @return [self] returns itself
+        def add_link(span_context_or_link_formatter, attrs = nil)
+          super
+          link =
+            if span_context_or_link_formatter.instance_of?(SpanContext)
+              OpenTelemetry::Trace::Link.new(span_context_or_link_formatter, attrs)
+            else
+              span_context_or_link_formatter.call
```

Comment:
The benefit comes about because a decision **not** to sample results in a no-op Span, which won't call the callable. For positive sampling decision, the Span would eventually be finished anyway, so deferring the call of the callable isn't terribly useful.

Reply:
I was thinking about the case where a link isn't used because max_links was reached, but this is probably not a huge concern and would fall into the premature optimization category.